server/src/test/java/org/elasticsearch/plugins/IndexStorePluginTests.java

```diff
@@ -92,4 +127,20 @@ public void testDuplicateIndexStoreFactories() {
         }
     }
 
+    public void testDuplicateIndexStoreRecoveryStateFactories() {
+        final Settings settings = Settings.builder().put("path.home", createTempDir()).build();
+        final IllegalStateException e = expectThrows(
+            IllegalStateException.class, () -> new MockNode(settings, Arrays.asList(FooCustomRecoveryStore.class,
+                                                                                    BarCustomRecoveryStore.class)));
+        if (JavaVersion.current().compareTo(JavaVersion.parse("9")) >= 0) {
+            assertThat(e, hasToString(matches(
+                "java.lang.IllegalStateException: Duplicate key recovery-type \\(attempted merging values " +
+                    "org.elasticsearch.plugins.IndexStorePluginTests\\$RecoveryFactory@[\\w\\d]+ " +
+                    "and org.elasticsearch.plugins.IndexStorePluginTests\\$RecoveryFactory@[\\w\\d]+\\)")));
+        } else {
+            assertThat(e, hasToString(matches(
+                "java.lang.IllegalStateException: Duplicate key " +
+                    "org.elasticsearch.plugins.IndexStorePluginTests$RecoveryFactory@[\\w\\d]+")));
```

Comment:
I think we shouldn't be asserting that our recovery factories have a `toString()` that returns `org.elasticsearch.plugins.IndexStorePluginTests$RecoveryFactory@[\\w\\d]+`: the fact that the message contains the duplicated key (and the string `Duplicate key`) should be enough IMO.

Bit of a pain the message doesn't contain the key in Java 8 but such is life.